compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -391,28 +391,45 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                            match (&ty.peel_refs().kind(), &actual.peel_refs().kind()) {
-                                (ty::Adt(def, _), ty::Adt(def_actual, _)) if def == def_actual => {
-                                    // Use `actual` as it will have more `substs` filled in.
-                                    self.ty_to_value_string(actual.peel_refs())
-                                }
-                                _ => self.ty_to_value_string(ty.peel_refs()),
+                    let (ty_str, placeholders) = if let Some(CandidateSource::Impl(impl_did)) =
+                        static_sources.get(0)
+                    {
+                        // When the "method" is resolved through dereferencing, we really want the
+                        // original type that has the associated function for accurate suggestions.
+                        // (#61411)
+                        let ty = tcx.at(span).type_of(*impl_did);
+                        match (&ty.peel_refs().kind(), &actual.peel_refs().kind()) {
+                            (ty::Adt(def, _), ty::Adt(def_actual, substs)) if def == def_actual => {
+                                // If there are any inferred arguments, (`{integer}`), we shouldn't mark
+                                // this as machine-applicable.
+                                let placeholders = substs
+                                    .iter()
+                                    .filter_map(|arg| {
+                                        if let GenericArgKind::Type(ty) = arg.unpack() {
+                                            Some(ty)
+                                        } else {
+                                            None
+                                        }
+                                    })
+                                    .any(|ty| matches!(ty.kind(), ty::Infer(_)));
+                                // Use `actual` as it will have more `substs` filled in.
+                                (self.ty_to_value_string(actual.peel_refs()), placeholders)
                             }
-                        } else {
-                            self.ty_to_value_string(actual.peel_refs())
-                        };
+                            _ => (self.ty_to_value_string(ty.peel_refs()), true),
+                        }
+                    } else {
+                        (self.ty_to_value_string(actual.peel_refs()), true)
+                    };
+                    let applicability = match placeholders {
+                        true => Applicability::HasPlaceholders,
+                        false => Applicability::MachineApplicable,
+                    };
                     if let SelfSource::MethodCall(expr) = source {
                         err.span_suggestion(
                             expr.span.to(span),
                             "use associated function syntax instead",
                             format!("{}::{}", ty_str, item_name),
```

Comment:
Instead of changing the applicability, we could replace unnameable / non-suggestable types with underscores (inference variables) allowing the diagnostic to stay machine-applicable. Similar to here:

https://github.com/rust-lang/rust/blob/a37499ae66ec5fc52a93d71493b78fb141c32f6b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs#L459-L480